servoshell: Replace `getopts` with `bpaf` for argument parsing

Narfinger opened the pull request:
This includes some small refactoring and some small breaking changes as
listed below. Other than these I tried to keep the functionality exactly
the same but because in the old code the parsing and settings of
preferences was intermingled it was difficult to figure out.

Small Breaking:
- Size and resources-path were unused but appeared in the help.
- soft-fail and hard-fail: Soft-fail flag got removed because it is too
  difficult to keep both. The default is now soft-fail and hard-fail can
be enabled.
- The help strings are obviously formatted differently now.
- -V does not work anymore but -v and --version.

Ideally, we want to have the ServoShellPreferences and Preferences be directly the Argument structure but that needs a bit more discussion because it would break backwards compatibility with the commandline.

This increases the binary size by ~280kb.

Testing: The testcases are still working but they do not cover much.
I added a unit test for the -p flag because it is the most difficult to parse in general.
Fixes: This will fix a small number of various parsing misshaps. It will also show if we are removing an option via unused lint.

mrobinson reviewed (changes requested):
Nice. A few initial comments, but I also need to give this a more thorough review. Thanks.

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
-        Size2D::new(components[0], components[1])
-    };
+    /// Wether or not to completely ignore certificate errors
+    #[bpaf(long, fallback(false))]
```
:information_source: FWIW `long` is the default argument name representation. You can keep them for documentation purposes (I usually do) but just having something like this will give the same result.

```rust
/// Wether or not to completely ignore certificate errors
ignore_certificate_errors: bool,
```

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+
+    /// Time profiler flag and either a TSV output filename OR and interval
+    /// for output to Stdout (blank for Stdout with interval of 5s)
+    #[bpaf(short('p'), argument::<String>("10 OR time.tsv"), parse(parse_profiling), fallback(None))]
```
The way it is written now results in a bit strange function signature for `parse_profiling` that always produces `Some`:

```rust
fn parse_profiling(arg: String) -> Result<Option<OutputOptions>, std::io::Error> {
```

You can change it to 

```rust
fn parse_profiling(arg: String) -> Result<OutputOptions, std::io::Error> {
```

and add `optional` [1] at the end of annotation:

```rust
    #[bpaf(short('p'), argument::<String>("10 OR time.tsv"), parse(parse_profiling), fallback(None))]
    profile: Option<OutputOptions>,
```

What's more if you can implement `FromStr` for `OutputOptions` you can let `argument` [2] to handle parsing with something like this:

```rust
    #[bpaf(short('p'), argument("10 OR time.tsv"))]
    profile: Option<OutputOptions>,
```

1. https://docs.rs/bpaf/latest/bpaf/trait.Parser.html#method.optional
2. https://docs.rs/bpaf/latest/bpaf/params/struct.NamedArg.html#method.argument

pacak replied:
Also `short` takes first character of a field name and uses it if you don't specify it. So it can be just this:
```rust
    #[bpaf(short, argument("10 OR time.tsv"))]
    profile: Option<OutputOptions>,
```

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    random_pipeline_closure_seed: Option<usize>,
+
+    /// Run in sandbox if multiprocess
+    #[bpaf(short('S'), long, fallback(true))]
```
`fallback(true)` won't do anything here. For inverted `switch` you can use `flag` [1] with arguments flipped.

1. https://docs.rs/bpaf/latest/bpaf/params/struct.NamedArg.html#method.flag

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    shaders: Option<PathBuf>,
+
+    /// Override the screen resolution in logical (device independent) pixels
+    #[bpaf(short('x'), long("screen-size"), argument::<String>("1024x768"),
```
Same as above you can use `FromStr` and `optional` to make code a bit more idiomatic.

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    user_agent: Option<String>,
+
+    /// Uses userscripts in resources/user-agent-js or a specified full path
+    #[bpaf(long, fallback(Some(PathBuf::from("resources/user-agent-js"))))]
```
This will always be `Some`, either value passed by user or value from a `fallback` You probably can drop `Option` part unless this is needed somewhere else.

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+
+    /// A user stylesheet ot be added to every document
+    #[bpaf(fallback(String::from("")), argument::<String>("path.css"), parse(parse_user_stylesheets))]
+    user_stylesheets: Vec<(Vec<u8>, ServoUrl)>,
```
I'd do something like this:

```rust
    #[bpaf(argument::<String>("path.css"), parse(parse_user_stylesheets), fallback(vec![]))]
   user_stylesheets: Vec<(Vec<u8>, ServoUrl)>,
```

It would avoid calling `parse_user_stylesheets` if none passed.

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    /// The url we should load
+    #[bpaf(
+        positional("URL"),
+        fallback(Some(String::from("https://www.servo.org")))
```
`Option` does nothing here - it will always be either value passed by user or `"servo.org"`.

pacak replied:
For now having `positional` items comes with a requirement that it must be at the end - I'd add a test to make sure this invariant isn't broken:

https://docs.rs/bpaf/latest/bpaf/struct.OptionParser.html#method.check_invariants

Narfinger replied:
I am actually having some problem with this. With the current version (slightly changed from the previous version you commented on), having `./target/release/servo github.com` returns `Error: `github.com` is not expected in this context. Any idea what this is about?

pacak replied:
On a first glance looks good. Can you paste output of `servo --help`?

Narfinger replied:
Still does not work. I checked that the args we are parsing is ["./target/release/servo", "www.github.com"]. That looks right but does run_inner perhaps expect something else?

pacak replied:
This is exactly what it expects. Let me try to build it myself...

Narfinger replied:
Ok I will try that when I find the time.

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +182,228 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    #[bpaf(fallback(String::from("")), argument::<String>("path.css"), parse(parse_user_stylesheets))]
+    user_stylesheets: Vec<(Vec<u8>, ServoUrl)>,
+
+    /// Prints the version information
```
There's `version` [1] that handles printing the version and exiting, you can customize the value you want to print as well.

1. https://docs.rs/bpaf/latest/bpaf/struct.OptionParser.html#method.version

Narfinger replied:
One thing which the documents do not state how to use a function for version with the derive. Currently we want to return the following combined string for version ` format!("Servo {}-{}", env!("CARGO_PKG_VERSION"), env!("GIT_SHA"))`.

pacak reviewed (commented):
Let me know if you have any problems with it :)

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -181,393 +183,233 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
-        preferences.devtools_server_port = port;
-    }
+    /// A preference to set
+    #[bpaf(argument("dom_bluetooth_enabled"),fallback(vec![]))]
```
You don't need `fallback(vec![])` here I think. If you want to be explicit - you can say  

```rust
#[bpaf(argument("dom_bluetooth_enabled"), many)]
pref: Vec<String>,
```

Same with `prefs_files` below.

pacak reviewed (commented):
Some more potential improvements for user experience:

Right now your "Usage" line contains all the things with full names - a few lines of text. You can remove less important bits with `hide_usage` [1] - items will still show up in the full help and will work as usual.

Right now everything goes into a single big product type which is a good start, but it might be useful to split that big structure into smaller sections like I'm doing in `cargo-show-asm`: [2]. Benefits to me as a developer is that it makes it easier to decide where to put a new option and decouples running and rendering parts - I know for sure that anything in cargo options shouldn't affect rendering. Benefits to me as a user is that in `--help` output I don't have to go though a wall of text and can figure out what section I want first and then look though that section.

1. https://docs.rs/bpaf/latest/bpaf/trait.Parser.html#method.hide_usage

2. https://github.com/pacak/cargo-show-asm/blob/master/src/opts.rs

mrobinson reviewed (changes requested):
Generally looks good, though I have a few requests:

Narfinger commented on ports/servoshell/prefs.rs:
```diff
@@ -191,402 +194,262 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    #[bpaf(argument("0"))]
+    devtools: Option<u16>,
+
+    /// Wether or not to enable experimental web platform features.
```
I went through all the strings again and copied them from the current version.
Yes, the rustdoc will be the help.

Narfinger committed:
Servoshell: Replace getopts with bpaf for parsing arguments.

This includes some small refactoring and some small breaking changes as
listed below. Other than these I tried to keep the functionality exactly
the same but because in the old code the parsing and settings of
preferences was intermingled it was difficult to figure out.

Small Breaking:
- Size and resources-path were unused but appeared in the help.
- soft-fail and hard-fail: Soft-fail flag got removed because it is too
  difficult to keep both. The default is now soft-fail and hard-fail can
be enabled.

Thanks to @pacak for all the help!

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -210,438 +213,333 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
     #[cfg(target_env = "ohos")]
-    let log_to_file = opt_match.opt_present("log-to-file");
+    /// Also log to a file (/data/app/el2/100/base/org.servo.servo/cache/servo.log).
+    #[bpaf(long, fallback(false))]
```
switch gets automagical fallback value of `false`

```suggestion
    #[bpaf(long)]
```

pacak commented on ports/servoshell/prefs.rs:
```diff
@@ -210,438 +213,333 @@ pub fn read_prefs_map(txt: &str) -> HashMap<String, PrefValue> {
+    random_pipeline_closure_seed: Option<usize>,
+
+    /// Run in a sandbox if multiprocess.
+    #[bpaf(short('S'), long, flag(true, false))]
```
```suggestion
    #[bpaf(short('S'), long)]
```

jdm commented:
What is the return code with --version?

pacak commented:
> What is the return code with `--version`?

Should be 0

Narfinger commented:
Let me check if I can debug this issue.

Narfinger commented:
@pacak So call on macos and linux I get return code of 1. I assume this is a bug?

Narfinger commented:
Ok it turns out that I misunderstood how some arugments should work. getopts::match_default returns None if the flag is not there but a default value if the flag is there without an argument.
The userscripts_directory should have defaulted to a None value when not given and not `resources/user-agent-js`. It did not matter because only in the macos image there were no directories to be found.

The latest changes this and also restores the original way webdriver port works. 
The big change in here is a new function that returns a parser `flag_with_default_parser` that replicates the behavior so we can use it for a couple of things.

pacak commented on ports/servoshell/lib.rs:
```diff
@@ -97,9 +97,7 @@ pub fn init_tracing(filter_directives: Option<&str>) {
     }
 }
 
-pub fn servo_version() -> String {
```
probably this

Narfinger commented:
Hopefully fixed. Let's see if I forgot any other targets.

pacak commented:
Lots of errors from the cli parser it seems.

```
  └ Error: `-u` is not expected in this context
```

 I don't see any obvious reasons for it to fail this way - it could be that `-u` was passed multiple times. Not sure what the inputs are...

Narfinger commented:
Clearly I just wanted to test if the merge queue works -.-
Let's try again.
